### ISSUE TITLE
Telemetry with ELRS 3.3.x

### ISSUE THREAD

**Issue by Target0815** [mstrens/expressLRS_relay#3]
Hi Michel,

I have tested the latest version (https://github.com/mstrens/expressLRS_relay/blob/test-without-wizio/Frsky_Elrs_Relay.uf2). 

**Relay operation with X9D+ => X8R => BetaFPV 2G4 TX with Firmware 3.3x**

![screen-2023-06-27-172338](https://github.com/mstrens/expressLRS_relay/assets/31627548/7290221d-1c28-4eed-8287-b6ef49d9898b)
![screen-2023-06-27-172351](https://github.com/mstrens/expressLRS_relay/assets/31627548/2e1db1fe-55a7-4958-bdc4-e0dd1bd56757)
![screen-2023-06-27-172403](https://github.com/mstrens/expressLRS_relay/assets/31627548/9415ac0e-d971-4d5c-b6d5-111e7f3f9365)
![screen-2023-06-27-172413](https://github.com/mstrens/expressLRS_relay/assets/31627548/3e4f751b-5435-43b0-a78c-5af1c16d0cb4)
![screen-2023-06-27-172425](https://github.com/mstrens/expressLRS_relay/assets/31627548/d654255e-2987-4484-9b46-b3c71c027e84)

Many "OC0?" values, implausible values like CURR etc.

**X9D+ => BetaFPV 2G4 TX with Firmware 3.3x**

![screen-2023-06-27-173205](https://github.com/mstrens/expressLRS_relay/assets/31627548/3ea6c54f-256a-4034-b943-e4c2446dff8b)
![screen-2023-06-27-173214](https://github.com/mstrens/expressLRS_relay/assets/31627548/ca264fb6-7bac-403b-9611-65a9e460766d)
![screen-2023-06-27-173223](https://github.com/mstrens/expressLRS_relay/assets/31627548/a477edd6-c5d0-44fc-b05d-5c0d384a2747)
![screen-2023-06-27-173234](https://github.com/mstrens/expressLRS_relay/assets/31627548/4c5a35f6-9713-4244-a416-5c3b9d532eed)

With the TX module in the module slot, the telemetry info fits.

**Comment by mstrens** [comment]
I do not find directly the bug and I have no hardware to test it myself.
So, I made some changes in order to add debugging messages.
I put this new version on github.
Could yo test it?
To do so, please:
- upload the new version on your RP2040
- connect the RP2040 to the FRsky receiver, to the ELRS TX and also to the PC (via the USB). Note: when RP2040 is connected to pc via USB, it is safe to avoid to power it on via another way (e.g. via the Frsky rx)
- start a serial terminal program on your PC (it can the serial terminal from the arduino IDE, from Vscode or any other like candle...). Select the good serial (COM) port. Set the baudrate to 115200

Your pc should now display 2 types of lines (when the chain FRsky TX - Frsky Rx - Relay - ELRS TX - ELSRS-RX - sensors is active):
-one line should give the content of each telemetry frames received by the RP2040 from the ELRS Tx module
- one line should give the index of each data sent by RP2040 via sport, the value to be sent and the content of the frame sent to the Frsky receiver.
Please post here an extract of some lines being generated. 
I hope it should help to find and fix the bug.

**Comment by mstrens** [comment]
I forgot to say that the new version is in the branch "test-without-wizio"

**Comment by Target0815** [comment]
The test log:

[relay_log.txt](https://github.com/mstrens/expressLRS_relay/files/11908742/relay_log.txt)

**Comment by mstrens** [comment]
Thanks.
I think I found the bug about the wrong value for the current.
The fields with an ID like OCxx are fields that refers to the quality of the link between ELRS TX and ELRS RX.
The relay get those data from ELRS but there is no predefined Frsky fields to transmit them over Sport. So I created special codes for them.
Here the list:
#define UPLINK_RSSI_1_ID        0x0c00  // to check if this range is valid
#define UPLINK_RSSI_2_ID        0x0c01
#define UPLINK_LINK_QUALITY_ID  0x0c02
#define UPLINK_SNR_ID           0x0c03
#define ACTIVE_ANTENNA_ID       0x0c04
#define RF_MODE_ID              0x0c05
#define UPLINK_TX_POWER_ID      0x0c06
#define DOWNLINK_RSSI_ID        0x0c07
#define DOWNLINK_LINK_QUALITY_ID  0x0c08
#define DOWNLINK_SNR_ID           0x0c09

Those codes does not have predefined name but I think that you can assign them a meaningful name in openTx/edge once they are discovered.

There are 2 other fields that ELRS transmit but that do no exist in Sport protocol. It is Number of sat and Capacity. Those 2 fields are put in frsky standard Temp1 and Temp2.

I put a new version on github (test-without-wizio branch). This version still contains the debug messages as in previous version. Let me know if it is OK and I can make a version where I remove those debug messages.

**Comment by Target0815** [comment]
Thanks!

I will do another test later.

**Comment by Target0815** [comment]
> I think I found the bug about the wrong value for the current.

No. Currently 70A are displayed, although it may be 700 mA. But I still have to measure or look at the OSD on the FC.

> Those codes does not have predefined name but I think that you can assign them a meaningful name in openTx/edge once they are discovered.

> There are 2 other fields that ELRS transmit but that do no exist in Sport protocol. It is Number of sat and Capacity. Those 2 fields are put in frsky standard Temp1 and Temp2.

Almost all data fields are defined in the FrSky Passthrough protocol ... no need for crutches like Tmp1/2. Have a look there:
S.Port Passthrough protocol specs: https://docs.google.com/spreadsheets/d/1oFv5zSl10wyR0LOcCChDdy9peIRNkbaEBrsHDQ0ZdmE/edit#gid=0

Passthrough over CRSF: https://github.com/yaapu/FrskyTelemetryScript/wiki/Passthrough-over-CRSF-and-ExpressLRS


I will do further tests later to see if the values are correct ...

**Comment by Target0815** [comment]
This already looks quite good ... 

GSpd is given in kts, should be km/h
Curr is 70A, in reality only 0,7A

Questions:
Tmp2 = Capacity = the capacity of the battery? I have here 12 as value, but can't assign it

0C00 to 0C09: UPLINK_* = Relay => Plane, DOWNLINK_* = Relay => Handset?

**Comment by mstrens** [comment]
I do not know how you analyse the data.
When I look at the specification of Sport, it seems that:
- GSpd as to be provided in 0.01 kts (ELRS transmits in 0.1 km/h and the relay converts to 0.01kts)
- current has to be provided in 0.001A (note: ELRS transmits it in 0.1A and the relay multiplies by 100 to get it in 0.001A)
I expect it is what is done by the relay project.
Capacity is what I get from ELRS. I expect it is the consumed capacity from the battery in mAh
UPLINK is related to the link between ELRS TX to ELRS RX
Downlink is related to the link between ELRS RX to ELRS TX

**Comment by Target0815** [comment]
> I do not know how you analyse the data.

I take the ELRS TX I use for the relay and put it directly into the transmitter and re-read the telemetry. The FC and the ELRS RX in the plane are always the same.

So I can compare very well which telemetry data is delivered via ELRS CRSF and what your converter makes out of it. And there you can see that Curr can't be right.

If I understood correctly, you don't have an ELRS system for testing? May I donate one to you, so that you do not have to handle the data blindly?

**Comment by mstrens** [comment]
Thanks for testing. The way you use is for sure the best one.
I checked the doc of EdgeTx once more and I saw that in Sport  Current is in 0.1A (and not in 0.001A as I expected).
So, there is no reason to multiply by 100 in the relay project.
I made the change in a new version I put on github now.

I got once an ELRS system that I used when I developed the code.
In the mean time, I does not have it anymore for testing.
So currently I make blind tests and corrections. I also do not have a flight controller to get data.
Usually I use only Frsky system (RX8) with openXsensor to get telemetry data. 
I am flying only F5J gliders.

About S.Port Passthrough protocol I would have to look further more to the specs.
It is not yet 100% clear for me.

**Comment by Target0815** [comment]
> So currently I make blind tests and corrections. I also do not have a flight controller to get data.

I consider the condition unsatisfactory ... I can provide you with ELRS TX and RX as well as a flight controller with ArduPlane, or donate the equivalent value. You just have to tell me how you want it.

> About S.Port Passthrough protocol I would have to look further more to the specs.
> It is not yet 100% clear for me.

Alessandro can explain this to you in detail: https://github.com/yaapu I am only a user but very satisfied.
